linalg/vector.cpp

```diff
@@ -142,11 +142,8 @@ Vector &Vector::operator=(const Vector &v)
 
 Vector &Vector::operator=(Vector &&v)
 {
-   data = std::move(v.data);
-   // Self-assignment-safe way to move v.size to size:
-   const auto size_tmp = v.size;
-   v.size = 0;
-   size = size_tmp;
+   v.Swap(*this);
```

Comment:
small question (outside of this PR):

The swap operation in `vector.hpp` is given by
```
inline void Vector::Swap(Vector &other)
{
   mfem::Swap(data, other.data);
   mfem::Swap(size, other.size);
}
```
The implementation of `Swap` is found in `array.hpp`:
```
template <class T>
inline void Swap(T &a, T &b)
{
   T c = a;
   a = b;
   b = c;
}
```

1. My current understanding is that this implementation would use copy assignment for `Memory`:
```
   /// Copy-assignment operator: default.
   Memory &operator=(const Memory &orig) = default;

   /** Move assignment operator. Sets the pointers and associated ownership of
       validity flags of @a *this to those of @a other. Resets @a other. */
   Memory &operator=(Memory &&orig)
```
2. Would it make sense to migrate migrate to [std::swap](https://en.cppreference.com/w/cpp/algorithm/swap) which supports move assignments ? Happy to look into this further if there is interest.

Reply:
Copying a memory object is cheap, it's similar to copying a pointer (there is no underlying memcpy happening).

(But yes, I think we should either use `std::move` in `mfem::Swap` or just call `std::swap`. I actually made that change in PR #4028).